inc/saimpls.h

```diff
@@ -81,6 +81,31 @@ typedef enum _sai_inseg_entry_attr_t
      */
     SAI_INSEG_ENTRY_ATTR_NEXT_HOP_ID,
 
+    /**
+     * @brief Use TC for a label.
+     *
+     * Associate TC using SAI_SWITCH_ATTR_QOS_EXP_TO_TC_MAP or SAI_PORT_ATTR_QOS_EXP_TO_TC_MAP.
+     * Mainly used for E-LSP kind of MPLS tunnels, where EXP infers both TC and COLOR
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default true
+     */
+    SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP,
```

Comment:
What is the use case for this flag? By default, if the application installs a QoS map, it is supposed to be used for every packet. So why would someone need to disable the QoS mapping later in the inseg entry and what is the expected behavior in this case?

Reply:
In L-LSP tunnel, MPLS labels carry traffic class
In E-LSP tunnel, EXP carry traffic class

This flag facilitate programming either E-LSP or L-LSP kind of labels in ILM. For E-LSP, infer TC from EXP with this flag set to true. For L-LSP, infer TC from label with this flag set to false.

I can remove it if we intend to support only E-LSP.

Reply:
Right, for E-LSP there is a QoS map attached to port or switch. For L-LSP, the `SAI_INSEG_ENTRY_ATTR_QOS_TC` is set, which **implies** **not to use** the QoS map but to set TC according to the label, which makes `SAI_INSEG_ENTRY_ATTR_USE_EXP_TO_TC_MAP` redundant.

Reply:
SAI_INSEG_ENTRY_ATTR_QOS_TC has default value zero. Perhaps it should be CREATE_ONLY and no default value? Giving default value mean it may always be set and always override QoS map.

Reply:
It can stay `CREATE_AND_SET`, but `@default disabled`.

Reply:
@default disabled doesn't work. Build fails saying (make in directory meta)

```
ERROR: invalid default value 'disabled' on SAI_INSEG_ENTRY_ATTR_QOS_TC (sai_uint8_t)
```
Without default it fails saying
```
 ASSERT FAILED (on line 445): SAI_INSEG_ENTRY_ATTR_QOS_TC: expected default value, but none provided
```
With CREATE_ONLY it fails
```
 ASSERT FAILED (on line 445): SAI_INSEG_ENTRY_ATTR_QOS_TC: expected default value, but none provided
```

I think we need boolean flag to make it "validonly" if L-LSP is desired. Otherwise some value is required. 

Traffic class is sai_cos_t or uint8_t, another way is to define invalid traffic class in saitypes.h

```
#define SAI_INVALID_TC 0xff
```

And use it as default value for it, but this I don't know is desirable.

Reply:
How about `@default internal`? I see it is used for many optional port and switch attributes.

Reply:
> How about `@default internal`? I see it is used for many optional port and switch attributes.

It produces error

```
 ASSERT FAILED (on line 1120): SAI_INSEG_ENTRY_ATTR_QOS_TC: default internal currently can be set only on read only objects
```

Reply:
I see. Then we probably should follow some examples of other optional attributes in SAI. For instance, SAI next-hop has an optional attribute VNI for tunnel encapsulation, which overrides the VNI mapping provided with the tunnel:

```
    /**
     * @brief Next hop entry VNI (override tunnel mapper)
     *
     * @type sai_uint32_t
     * @flags CREATE_AND_SET
     * @default 0
     * @validonly SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_TUNNEL_ENCAP
     */
    SAI_NEXT_HOP_ATTR_TUNNEL_VNI,
```

It has a `@default 0` tag, but what it really means is that if the attribute is provided, then use it, otherwise just use the tunnel mapping, but not use value 0. It doesn't have any explicit attribute like `SAI_NEXT_HOP_ID_USE_TUNNEL_VNI_MAP`. SAI driver implementation makes the decision to override the mapping by the **presence** of this attribute in the `create_next_hop()` call.

So I think we should follow the same path having annotation `@default 0` and just mention in the comment of the `SAI_INSEG_ENTRY_ATTR_QOS_TC` attribute that it overrides EXP to TC mapping, and remove a flag.

Reply:
makes sense, if attribute is present only then it overrides even with default value. If not present, its default value doesn't take an effect.